### ISSUE TITLE
Terraform plan should not lock state files by default since it no longer writes changes to state

### ISSUE THREAD

**Issue by davidalger** [hashicorp/terraform#28130]
### Terraform Version
```
0.14.8
```

### Expected Behavior

`terraform plan` can be run simultaneously by multiple developers and parallel CI jobs such as will happen if pull/merge requests run `terraform plan` every time they are updated, prior to being merged, etc.

### Actual Behavior

```
Error: Error locking state: Error acquiring the state lock: writing "gs://<redacted>/default.tflock" failed: googleapi: Error 412: Precondition Failed, conditionNotMet
Lock Info:
  ID:        1616009873855711
  Path:      gs://<redacted>/default.tflock
  Operation: OperationTypePlan
  Who:       <redacted>
  Version:   0.14.8
  Created:   2021-03-17 19:37:53.540992 +0000 UTC
  Info:      


Terraform acquires a state lock to protect the state from being written
by multiple users at the same time. Please resolve the issue above and try
again. For most commands, you can disable locking with the "-lock=false"
flag, but this is not recommended.
```

### Steps to Reproduce

Run `terraform plan` on the same root module (with the same state) twice at the same time.

### Additional Context

In an earlier time, running `terraform plan` would refresh state _and_ update remote state with those changes incrementing the serial number. As a result, state locking for `plan` was a hard requirement, otherwise one could experience an operation overwriting the state updates of another. Fast-forward to where we are today and the refresh `terraform plan` does is merely held in-memory, yet the state file is still locked for the duration of the plan operation. With 0.14 the entire in-memory refresh was even overhauled with a lot of work being done to improve the data lifecycle during plan operations (#26270, #26285, #26321) such that the refresh phase no longer even occurs, as resources are updated on-demand during planning. This in-memory refreshed state is used to generate the plan, so there should be no further communication with the remote state following an initial read.

To quote Terraform documentation (or the error message above), "Terraform will lock your state for all operations that could write state. This prevents others from acquiring the lock and potentially corrupting your state."

In this case however, plan should no longer be making any writes to state making the lock relatively useless, as the entire state file should be loaded in a single atomic read operation at the start of the plan cycle. It's been so long since Terraform wrote changes to state I've forgotten when exactly it stopped, but it was long before recent versions of Terraform existed.

I'm not the only one to ask this question, see [Why does a plan lock the state file?](https://discuss.hashicorp.com/t/why-does-a-plan-lock-the-state-file/3520) on the community forum, but I cannot find any good reason why `terraform plan` continues to lock the state file to this present day. If there is a good reason for this, would some documentation be in order?

There is a workaround, yes, I can specify `-lock=false` and be done with it as far as CI pipelines go. That's tough for a developer to remember to type out, and being overly cautious many simply won't as a result of the warning not to use it.

What I'd like to propose is changing the `-lock` flag to default to `false` when `terraform plan` is run, deprecating it's use for `plan` operations, eventually removing the flag entirely after a sufficient deprecation timeline has elapsed.

**Comment by jbardin** [comment]
Hi @davidalger,

Thanks for filing the request. Just to provide a quick answer for why we can't immediately implement this; the way the remote state backend code is structured, remote state implementations can still write to their underlying storage when requesting access to a state, even if it's not writing the terraform state itself. For this reason, internally when a locking is possible, the implementation must take the lock to update the storage structures.

Another type of consideration is that if an apply is happening concurrently to the plan operation, the plan may show partially or fully invalid results. Now if this plan is saved, it could of course not be applied since the state would not match the new changes, but the result would be confusing to users who are assuming only one operation can happen at a time.

We are however working on restructuring remote state storage in general, and specifying all the requirements and details around locking of state storage will be part of that project. Once the new specification for remote state locking is defined, it will likely guide any changes to how state is locked in the CLI.

**Comment by etaham** [comment]
@jbardin 

> Another type of consideration is that if an apply is happening concurrently to the plan operation, the plan may show partially or fully invalid results. Now if this plan is saved, it could of course not be applied since the state would not match the new changes, but the result would be confusing to users who are assuming only one operation can happen at a time.

Could a no-lock solution be employed here - after running the plan, compare the version/hash of the state downloaded to the current state. If the version/hash is the same, then you're good. If not, either warn or automatically rerun the plan with the new state. It could also warn if it detects that the state is locked that a new plan may be needed (or wait and auto-retry the plan there too).

**Comment by mwos-sl** [comment]
This kind of locking problems are well known in multi-thread world (this one is known as [multiple reader single write](https://en.wikipedia.org/wiki/Readers%E2%80%93writer_lock)) and it is solvable.
TLDR is: allow multiple readers (`terraform plan`) to read if no one is writing, and allow single writer to write (`terraform apply` and the others that changes the state) only when no one reads).

**Comment by chefren** [comment]
This is in the [docs](https://www.terraform.io/language/state/locking) (not sure since when)
```You can disable state locking for most commands with the -lock flag but it is not recommended.```

@jbardin is it safe to say that if the use case is expected "pure read only and not too concerned about changing state or aware of changing state" this flag covers it? Or would `-lock=false` also have other problems apart from the previously mentioned?

**Comment by drdamour** [comment]
> HI @etaham, that leaves terraform in the same position it is currently, as an apply operation could still be in progress, and finish immediately after the plan completes, so there is little benefit to the extra operations. Any change here is really going to be dependent on a new remote storage design, and whatever locking semantics arise from that design. The different capabilities of various backends makes it difficult to have a generalized locking solution that suits everyone, since implementing different levels of locking may not be feasible for all services.

so why can't this be a flag in the backend statement then for each provider.  `default_plan_lock = false` since it would work fine in most of the backends people are using like aws & azure.

**Comment by rem-aj** [comment]
@davidalger  or anyone currently using the workaround of not locking state on plan?
And have they run into the concerns @jbardin mentioned above?

Especially with regards to an apply running at same time as plan leading to an inconsistent saved plan?

**Comment by pavitra-infocusp** [comment]
> This kind of locking problems are well known in multiple-threads world (this one is known as [multiple reader single write](https://en.wikipedia.org/wiki/Readers%E2%80%93writer_lock)) and it is solvable. TLDR is: allow multiple readers (`terraform plan`) to read if no one is writing, and allow single writer to write (`terraform apply`, etc..) only when no one reads and write).
> 
> It would be great to have something like this in terraform.

Exactly my thoughts! Read-Write locks are widely used in this scenario.
